src/sentry.js

```diff
@@ -45,6 +52,86 @@ const preventNoise = (): void => {
   }
 };
 
+/**
+ * An HTTP request made by the app.
+ *
+ * One of the builtin "recognized breadcrumb types":
+ *   https://develop.sentry.dev/sdk/event-payloads/breadcrumbs/#breadcrumb-types
+ */
+type HttpBreadcrumb = $ReadOnly<{|
+  ...Breadcrumb,
+  type: 'http',
+  data: {|
+    url?: string,
+    method?: string,
+    status_code?: number,
+    reason?: string,
+  |},
+|}>;
+
+function shouldScrubHost(url: URL, accountStatuses: $ReadOnlyArray<AccountStatus>) {
+  if (isAppOwnDomain(url)) {
+    return false;
+  }
+
+  if (url.pathname.startsWith('/api/v1')) {
+    // Most likely an API request to a Zulip realm. This will catch
+    // `/api/v1/server_settings` and `/api/v1/fetch_api_key` requests to a
+    // realm before the account is added to the account-statuses state.
+    return true;
+  }
+
+  if (accountStatuses.some(({ realm }) => url.origin === realm.origin)) {
+    // Definitely a request to a Zulip realm. Will catch requests to realms
+    // in the account-statuses state, including those without `/api/v1`,
+    // like `/avatar/{user_id}` (zulip/zulip@0f9970fd3 confirms that this
+    // and potentially others don't / won't use `/api/v1`).
+    return true;
+  }
+
+  return false;
+}
+
+/**
+ * Scrubs possibly personal information from a breadcrumb.
+ *
+ * - Removes the realm hostnames in API requests, except for Kandra-hosted
+ *   realms.
+ */
+function scrubBreadcrumb(breadcrumb: Breadcrumb, hint?: BreadcrumbHint): Breadcrumb {
+  switch (breadcrumb.type) {
+    case 'http': {
+      // $FlowIgnore[incompatible-indexer] | We assume it's an
+      // $FlowIgnore[incompatible-type]    | HttpBreadcrumb; see jsdoc.
+      const httpBreadcrumb: HttpBreadcrumb = breadcrumb;
+
+      const unscrubbedUrl = httpBreadcrumb.data.url;
+      let scrubbedUrl;
```

Comment:
(Also I think actually both of these can be made somewhat cleaner by using control flow with `return`, instead of data flow mutating a `let`-bound variable. Maybe I'll try doing that at or after merging.)

Reply:
(Indeed I did; take a look.)

Reply:
e62493475